sfge/src/main/java/com/salesforce/rules/usewithsharingondatabaseoperation/UseWithSharingOnDatabaseOperationVisitor.java

```diff
@@ -0,0 +1,446 @@
+import com.salesforce.graph.visitor.DefaultNoOpPathVertexVisitor;
+import com.salesforce.rules.Violation;
+import com.salesforce.rules.avoiddatabaseoperationinloop.DatabaseOperationUtil;
+import com.salesforce.rules.ops.boundary.SharingPolicyBoundary;
+import com.salesforce.rules.ops.boundary.SharingPolicyBoundaryDetector;
+import java.util.HashSet;
+import java.util.Set;
+
+public class UseWithSharingOnDatabaseOperationVisitor extends DefaultNoOpPathVertexVisitor {
+
+    /** record if warnings are enabled or disabled */
+    private final boolean IS_WARNING_VIOLATION_DISABLED =
+            SfgeConfigProvider.get().isWarningViolationDisabled();
+
+    /** keeps track of the {@link SharingPolicyBoundary}s */
+    private final SharingPolicyBoundaryDetector sharingPolicyBoundaryDetector;
+
+    /** represents the path entry point that this visitor is walking */
+    private final SFVertex sourceVertex;
+
+    /**
+     * represents the database operation statement that is possibly inside an unsafe sharing policy
+     */
+    private final BaseSFVertex sinkVertex;
+
+    /** collects violation information */
+    private final HashSet<Violation.PathBasedRuleViolation> violations;
+
+    /**
+     * visits vertices along a {@link com.salesforce.graph.ApexPath} to determine if the sink
+     * vertex, a database operation, is executed from a "with sharing" policy.
+     *
+     * @param boundaryDetector a {@link SharingPolicyBoundaryDetector#getEffectiveSharingBoundary()}
+     *     that keeps track of the current sharing policy
+     * @param sourceVertex a {@link BaseSFVertex} that is the entry point of the path containing the
+     *     sink vertex
+     * @param sinkVertex a vertex that is a database opeartion. See {@link
+     *     DatabaseOperationUtil#isDatabaseOperation(BaseSFVertex)}
+     */
+    UseWithSharingOnDatabaseOperationVisitor(
+            SharingPolicyBoundaryDetector boundaryDetector,
+            BaseSFVertex sourceVertex,
+            BaseSFVertex sinkVertex) {
+        if (!(DatabaseOperationUtil.isDatabaseOperation(sinkVertex))) {
+            throw new ProgrammingException(
+                    "UseWithSharingOnDatabaseOperation Sink vertex must be a "
+                            + "MethodCallExpressionVertex in the Database class, DmlStatementVertex, or "
+                            + "SoqlExpressionVertex. Provided sink vertex="
+                            + sinkVertex);
+        }
+        this.sharingPolicyBoundaryDetector = boundaryDetector;
+        this.sourceVertex = sourceVertex;
+        this.sinkVertex = sinkVertex;
+        this.violations = new HashSet<>();
+    }
+
+    @Override
+    public void afterVisit(MethodCallExpressionVertex vertex, SymbolProvider symbols) {
+        // if this method call is also a database operation, handle it
+        if (DatabaseOperationUtil.isDatabaseOperation(vertex)) {
+            createViolationIfDatabaseOpUnsafe(vertex, symbols);
+        }
```

Comment:
Question: Is this check necessary? Doesn't the handler's `test` method prevent this from running against non-DB operation vertices?

Reply:
True!